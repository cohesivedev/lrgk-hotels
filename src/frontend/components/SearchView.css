.ResultTable {
    margin-top: 1em;
    width: 100%;
}

.ResultTable .Count {
    padding: 1em;
    color:white;
    background-color: #444444;
}

.ResultTable tr:nth-child(2n) {
    background-color: #eeeeee;
}

.ResultTable td {
    padding: 1em;
    vertical-align: top;
}

