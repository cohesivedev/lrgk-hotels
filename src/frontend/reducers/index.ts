import {Action} from 'redux';

import {IAppState} from '../store';
import resultsReducer from './results';

export default function reducers(state: IAppState, action: Action): IAppState {
    return {
        ...state,
        results: resultsReducer(state.results, action)
    };
}
