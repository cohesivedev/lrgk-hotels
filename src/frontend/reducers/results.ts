import {IResultsState} from '../store';
import {ResultsAction, ResultsActionTypes} from '../actions/results';

export default function (state: IResultsState, action: ResultsAction): IResultsState {
    let resultsState = {...state};

    switch (action.type) {
        case ResultsActionTypes.GET_RESULTS_HOTELS:
            const {payload} = action;


            resultsState = {
                ...resultsState,
                hotels:       payload!,
                validFilters: [] // todo: put all valid facility filters in here, computed and stored once per fetch
            };
            break;

        default:
    }

    return resultsState;
}
