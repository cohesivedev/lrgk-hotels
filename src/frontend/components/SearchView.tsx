import * as React from 'react';
import {ChangeEventHandler} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import {IAppState, IHotel} from '../store';
import {getHotels} from '../actions/results';
import {byAscendingStarRating, byDescendingStarRating, isSuperset} from '../helpers';

import RatingSorter from './RatingSorter';
import ResultItem from './ResultItem';
import FacilitiesFilter from './FacilitiesFilter';

interface ISearchViewProps {
    hotels: IHotel[];
    validFilters: string[];
}

interface ISearchViewStateProps {
    getHotels: Function;
}

interface ISearchViewState {
    isSortAscending: boolean;
    filters: string[];
}

class SearchView extends React.Component<ISearchViewProps & ISearchViewStateProps> {
    state: ISearchViewState = {
        isSortAscending: true,
        filters:         []
    };

    onChangeFilters: ChangeEventHandler<HTMLInputElement> = e => {
        const filter           = e.currentTarget.getAttribute('data-filter')!;
        const shouldBeIncluded = e.currentTarget.checked;

        const filters = [...this.state.filters];

        if (shouldBeIncluded) {
            filters.push(filter);
        } else {
            filters.splice(filters.indexOf(filter), 1);
        }

        this.setState({filters});
    };

    onChangeRatingSort: ChangeEventHandler<HTMLSelectElement> = e => {
        this.setState({isSortAscending: e.currentTarget.value === 'ascending'});
    };

    componentWillMount(): void {
        const {hotels, getHotels} = this.props;

        if (hotels.length === 0) {
            getHotels();
        }
    }

    render(): React.ReactNode {
        const {validFilters, hotels}     = this.props;
        const {filters, isSortAscending} = this.state;

        const filtersSet = new Set(filters);

        const results = hotels
            .filter(h => isSuperset(new Set(h.facilities), filtersSet))
            .sort(isSortAscending ? byAscendingStarRating : byDescendingStarRating)
            .map(ResultItem);

        return <>
            <FacilitiesFilter
                validFilters={validFilters}
                activeFilters={filters}
                onChange={this.onChangeFilters}
            />
            <RatingSorter onChange={this.onChangeRatingSort} isAscending={isSortAscending}/>
            <table>
                <tbody>
                <tr>
                    <th>Name</th>
                    <th>Star rating</th>
                    <th>Facilities</th>
                </tr>
                {results || 'No results found.'}
                </tbody>
            </table>
        </>
    }
}

const mapDispatchToProps = dispatch => (
    bindActionCreators({
        getHotels
    }, dispatch)
);

export default connect<ISearchViewProps, ISearchViewStateProps, null>(
    ({results}: IAppState) => ({
        hotels:       results.hotels,
        validFilters: results.validFilters
    }),
    mapDispatchToProps
)(SearchView);
