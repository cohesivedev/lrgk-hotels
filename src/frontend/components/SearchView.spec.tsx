import * as React from 'react';
import {renderToString} from 'react-dom/server';
import {Provider} from 'react-redux';
import {createStore} from 'redux';

import reducers from '../reducers';
import SearchView from './SearchView';

describe('components/SearchView', () => {
    it('should match snapshot', () => {
        const fakeStore = createStore(
            reducers,
            {
                results: {
                    hotels:       [
                        {name: 'foo', starRating: 5, facilities: ['a']},
                        {name: 'bar', starRating: 2, facilities: ['b', 'c']}
                    ],
                    validFilters: []
                }
            }
        );

        expect(renderToString(
            <Provider store={fakeStore}>
                <SearchView/>
            </Provider>
        )).toMatchSnapshot();
    });
});
