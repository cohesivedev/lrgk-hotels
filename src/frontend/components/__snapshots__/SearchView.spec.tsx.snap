// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SearchView should match snapshot 1`] = `

<div>
</div>
<div>
  <label for="rating-sorter">
    Sort star rating by
    <select id="rating-sorter">
      <option selected
              value="ascending"
      >
        Low to High (Ascending)
      </option>
      <option value="descending">
        High to Low (Descending)
      </option>
    </select>
  </label>
</div>
<table>
  <tbody>
    <tr>
      <th>
        Name
      </th>
      <th>
        Star rating
      </th>
      <th>
        Facilities
      </th>
    </tr>
    <tr>
      <td>
        bar
      </td>
      <td>
        2
      </td>
      <td>
        <div>
          b
        </div>
        <div>
          c
        </div>
      </td>
    </tr>
    <tr>
      <td>
        foo
      </td>
      <td>
        5
      </td>
      <td>
        <div>
          a
        </div>
      </td>
    </tr>
  </tbody>
</table>

`;
